File: tasks/native-runtime/tool_dispatch_parity_probe.py
"""Tool-dispatch parity probe, Python side.

Measures the *seam* of ``deepseek_infra/infra/tool_runtime/tools.py``: the
argument/limit/name normalization, the chart branch, and the envelope/ordering of
``execute_tool_call``.

What is extracted verbatim
--------------------------
``tool_call_name``, ``parse_tool_arguments``, ``safe_limit``,
``is_parallel_safe_tool``, ``generate_chart``, ``chart_markdown_table``,
``execute_tool_call``, ``SERIAL_TOOL_NAMES`` — plus the real ``ErrorCode`` /
``AppError`` by executing the self-contained ``core/errors.py``, and the real
``ToolPolicy`` / ``tool_metadata`` by reusing the tool-policy probe's namespace.

What is stubbed, and why
------------------------
- **The 18 branch functions.** Only ``generate_chart`` is pure enough to port
  without its packages, so the others are replaced by recorders. That is not a
  loss of fidelity for what is being measured here: it is what makes the *gate
  short-circuit* observable (a denied call must invoke **no** branch).
- **``_evaluate_rust_policy`` returns ``None``.** This is exactly what the real
  function does when ``DEEPSEEK_RUST_POLICY`` is false, which is the production
  default (``infra/rust_core/config.py``). The Rust-policy path is measured
  separately by the tool-policy probe.
- **``schema_for_tool`` returns ``None``.** The schema catalog derives from
  ``available_tool_definitions()``; schema-driven verdicts are covered by the
  tool-policy probe's ``validate::*`` keys instead of being duplicated here.

Nothing is written anywhere.

Usage::

    python tasks/native-runtime/tool_dispatch_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example tool_dispatch_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import ast
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import tool_policy_parity_probe as policy_probe  # noqa: E402

REPO = Path(__file__).resolve().parents[2]
TOOLS = REPO / "deepseek_infra" / "infra" / "tool_runtime" / "tools.py"
ERRORS = REPO / "deepseek_infra" / "core" / "errors.py"

EXTRACT_NAMES = (
    "tool_call_name",
    "parse_tool_arguments",
    "safe_limit",
    "is_parallel_safe_tool",
    "generate_chart",
    "chart_markdown_table",
    "execute_tool_call",
)

# Recorded branch invocations, so a gate short-circuit is observable.
CALLS: list[str] = []


def _extract_function(source: str, name: str) -> str | None:
    for node in ast.walk(ast.parse(source)):
        if isinstance(node, ast.FunctionDef) and node.name == name:
            return ast.get_source_segment(source, node)
    return None


def _extract_assignment(source: str, name: str) -> object:
    for node in ast.parse(source).body:
        if isinstance(node, ast.Assign):
            for target in node.targets:
                if isinstance(target, ast.Name) and target.id == name:
                    return ast.literal_eval(node.value)
    raise SystemExit(f"could not extract {name}")


def build_namespace() -> dict:
    tools_source = TOOLS.read_text(encoding="utf-8")
    namespace: dict = {}

    # The real error types — `errors.py` is self-contained.
    exec(compile(ERRORS.read_text(encoding="utf-8"), str(ERRORS), "exec"), namespace)  # noqa: S102

    from typing import Any, Callable  # noqa: E402
    import re  # noqa: E402
    import threading  # noqa: E402

    # The extracted functions reference these as module globals.
    namespace.update(
        {"Any": Any, "Callable": Callable, "json": json, "re": re, "threading": threading}
    )

    # The real policy objects, reused from the tool-policy probe. Audit is turned
    # off here: this probe measures dispatch, not the audit log, and the log path
    # is unset in that namespace.
    policy_ns = policy_probe.build_namespace()
    policy_ns["TOOL_POLICY_AUDIT_ENABLED"] = False
    for key in ("ToolPolicy", "tool_metadata", "PolicyDecision", "DENY"):
        namespace[key] = policy_ns[key]

    namespace["SERIAL_TOOL_NAMES"] = _extract_assignment(tools_source, "SERIAL_TOOL_NAMES")

    for name in EXTRACT_NAMES:
        found = _extract_function(tools_source, name)
        if found is None:
            raise SystemExit(f"could not extract {name}")
        exec(compile(found, str(TOOLS), "exec"), namespace)  # noqa: S102

    def recorder(branch: str):
        def stub(*_args: object, **_kwargs: object) -> dict:
            CALLS.append(branch)
            return {"stub": branch}

        return stub

    for branch in (
        "python_eval",
        "search_files",
        "fetch_url",
        "compare_search_results",
        "build_memory_suggestion",
        "create_reminder_tool",
        "list_reminders_tool",
        "recall_memory_tool",
        "forget_memory_tool",
        "list_project_files_tool",
        "read_file_chunk_tool",
        "data_transform",
        "create_mindmap",
        "create_presentation",
        "create_document",
        "browser_action_tool",
        "normalize_memory_scope",
        "call_external_mcp_tool",
    ):
        namespace[branch] = recorder(branch)

    # Mirrors the production default (`DEEPSEEK_RUST_POLICY` false) and the
    # schema source duplicating the tool-policy probe.
    namespace["rust_policy_enabled"] = lambda: False
    namespace["_evaluate_rust_policy"] = lambda *a, **k: None
    namespace["schema_for_tool"] = lambda name: None

    return namespace


# --- corpus ----------------------------------------------------------------------

PARSE_CASES: list[tuple[str, object]] = [
    ("object", {"a": 1}),
    ("object-string", '{"a": 1}'),
    ("nested-string", '{"a": {"b": [1, 2]}}'),
    ("not-json", "not json"),
    ("json-array", "[1, 2]"),
    ("json-scalar", "42"),
    ("blank", "   "),
    ("empty", ""),
    ("none", None),
    ("integer", 7),
]

LIMIT_CASES: list[tuple[str, object]] = [
    ("in-range", 3),
    ("zero", 0),
    ("negative", -9),
    ("above-max", 99),
    ("numeric-string", "7"),
    ("float-string", "7.9"),
    ("float", 3.9),
    ("unparseable", "abc"),
    ("none", None),
    ("list", [1]),
    ("bool-true", True),
]

NAME_CASES: list[tuple[str, dict]] = [
    ("function-name", {"function": {"name": " search_files "}}),
    ("blank-function-top-level", {"function": {"name": ""}, "name": "x"}),
    ("top-level-only", {"name": "y"}),
    ("missing", {}),
    ("function-not-a-dict", {"function": "nope", "name": "z"}),
]

PARALLEL_CASES: list[tuple[str, dict]] = [
    ("serial-tool", {"function": {"name": "web_search"}}),
    ("parallel-tool", {"function": {"name": "generate_chart"}}),
    ("nameless", {}),
]

CHART_CASES: list[tuple[str, dict]] = [
    ("line-with-title", {"type": "line", "title": "Revenue", "data": [{"label": "Q1", "value": 1}, {"label": "Q2", "value": 2.5}]}),
    ("defaults", {"data": [{"label": "a", "value": 1}]}),
    ("unknown-type", {"type": "radar", "data": [{"label": "a", "value": 1}]}),
    ("string-numbers", {"data": [{"label": "a", "value": "3"}]}),
    ("drops-and-caps", {"data": ["nope", {"value": 1}, {"label": "nv"}, {"label": "n", "value": None}, {"label": "bad", "value": "abc"}, {"label": "ok", "value": "3"}] + [{"label": f"p{i}", "value": i} for i in range(20)]}),
    ("pipe-in-label", {"data": [{"label": "a|b", "value": 1}]}),
    ("empty-list", {"data": []}),
    ("not-a-list", {"data": "nope"}),
    ("no-valid-points", {"data": [{"value": 1}]}),
    ("float-rendering", {"data": [{"label": "x", "value": 100}, {"label": "y", "value": -0.0}]}),
]

# (label, tool_call, policy mode) — `policy` is "none" or "permissive".
DISPATCH_CASES: list[tuple[str, dict, str]] = [
    ("unknown-no-policy", {"function": {"name": "not_a_tool", "arguments": "{}"}}, "none"),
    ("unknown-with-policy", {"function": {"name": "not_a_tool", "arguments": "{}"}}, "permissive"),
    ("chart-no-policy", {"function": {"name": "generate_chart", "arguments": '{"data": [{"label": "a", "value": 1}]}'}}, "none"),
    ("chart-with-policy", {"function": {"name": "generate_chart", "arguments": '{"data": [{"label": "a", "value": 1}]}'}}, "permissive"),
    ("chart-empty-args", {"function": {"name": "generate_chart", "arguments": "{}"}}, "none"),
    ("chart-arguments-as-object", {"function": {"name": "generate_chart", "arguments": {"data": [{"label": "a", "value": 2}]}}}, "none"),
    # Denied by the gate, so the branch must never be invoked.
    ("ssrf-denied-with-policy", {"function": {"name": "fetch_url", "arguments": '{"url": "http://169.254.169.254/"}'}}, "permissive"),
    ("path-denied-with-policy", {"function": {"name": "search_files", "arguments": '{"path": "../../etc/passwd"}'}}, "permissive"),
]


def _policy(namespace: dict, mode: str):
    if mode == "none":
        return None
    return namespace["ToolPolicy"].permissive()


def main() -> int:
    for path in (TOOLS, ERRORS):
        if not path.exists():
            print(f"missing {path}", file=sys.stderr)
            return 2

    namespace = build_namespace()
    out: dict = {}

    parse_tool_arguments = namespace["parse_tool_arguments"]
    safe_limit = namespace["safe_limit"]
    tool_call_name = namespace["tool_call_name"]
    is_parallel_safe_tool = namespace["is_parallel_safe_tool"]
    generate_chart = namespace["generate_chart"]
    chart_markdown_table = namespace["chart_markdown_table"]
    execute_tool_call = namespace["execute_tool_call"]
    AppError = namespace["AppError"]

    for label, value in PARSE_CASES:
        out[f"parse::{label}"] = parse_tool_arguments(value)

    for label, value in LIMIT_CASES:
        out[f"limit::{label}"] = safe_limit(value, default=5, maximum=10)

    for label, call in NAME_CASES:
        out[f"name::{label}"] = tool_call_name(call)

    for label, call in PARALLEL_CASES:
        out[f"parallel::{label}"] = is_parallel_safe_tool(call)

    for label, arguments in CHART_CASES:
        try:
            out[f"chart::{label}"] = {"ok": True, "result": generate_chart(
                str(arguments.get("type") or "bar"),
                str(arguments.get("title") or ""),
                arguments.get("data"),
            )}
        except AppError as exc:
            out[f"chart::{label}"] = {"ok": False, "error": str(exc), "code": exc.code.value}

    out["markdown::table"] = chart_markdown_table(
        [{"label": "a|b", "value": 1.0}, {"label": "c", "value": 2.5}]
    )
    out["serial::names"] = sorted(namespace["SERIAL_TOOL_NAMES"])

    for label, call, mode in DISPATCH_CASES:
        CALLS.clear()
        output = execute_tool_call(call, policy=_policy(namespace, mode))
        if output.get("ok") is True:
            outcome = "executed"
        elif output.get("code") == "forbidden":
            outcome = "denied"
        else:
            outcome = "unsupported"
        out[f"dispatch::{label}"] = {
            "outcome": outcome,
            "output": output,
            "branches": sorted(set(CALLS)),
        }

    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/tool_dispatch_parity_probe.rs
//! Tool-dispatch parity probe, Rust side.
//!
//! Replays the same fixed corpus as
//! `tasks/native-runtime/tool_dispatch_parity_probe.py` through
//! `deepseek_policy::tool_dispatch` and prints canonical JSON, so the two outputs
//! can be diffed byte-for-byte.
//!
//! Only the branches that exist on both sides are compared. [DispatchOutcome
//! ::Unported] deliberately has no envelope, so a branch that has not been ported
//! cannot be made to look like one that ran — those are covered by unit tests
//! instead (see `exactly_one_branch_is_ported_and_every_other_names_its_blocker`
//! and `an_unported_branch_has_no_envelope_at_all`).
//!
//! Usage::
//!
//!     python tasks/native-runtime/tool_dispatch_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example tool_dispatch_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use deepseek_policy::tool_batch::execute_tool_calls;
use deepseek_policy::tool_dispatch::{
    DispatchOutcome, SERIAL_TOOL_NAMES, chart_markdown_table, dispatch, generate_chart,
    is_parallel_safe_tool, parse_tool_arguments, safe_limit, tool_call_name,
};
use deepseek_policy::tool_policy::ToolPolicy;
use deepseek_policy::tool_transform::data_transform;
use serde_json::{Map, Value, json};

fn parse_cases() -> Vec<(&'static str, Value)> {
    vec![
        ("object", json!({"a": 1})),
        ("object-string", json!("{\"a\": 1}")),
        ("nested-string", json!("{\"a\": {\"b\": [1, 2]}}")),
        ("not-json", json!("not json")),
        ("json-array", json!("[1, 2]")),
        ("json-scalar", json!("42")),
        ("blank", json!("   ")),
        ("empty", json!("")),
        ("none", Value::Null),
        ("integer", json!(7)),
    ]
}

fn limit_cases() -> Vec<(&'static str, Value)> {
    vec![
        ("in-range", json!(3)),
        ("zero", json!(0)),
        ("negative", json!(-9)),
        ("above-max", json!(99)),
        ("numeric-string", json!("7")),
        ("float-string", json!("7.9")),
        ("float", json!(3.9)),
        ("unparseable", json!("abc")),
        ("none", Value::Null),
        ("list", json!([1])),
        ("bool-true", json!(true)),
    ]
}

fn name_cases() -> Vec<(&'static str, Value)> {
    vec![
        (
            "function-name",
            json!({"function": {"name": " search_files "}}),
        ),
        (
            "blank-function-top-level",
            json!({"function": {"name": ""}, "name": "x"}),
        ),
        ("top-level-only", json!({"name": "y"})),
        ("missing", json!({})),
        (
            "function-not-a-dict",
            json!({"function": "nope", "name": "z"}),
        ),
    ]
}

fn parallel_cases() -> Vec<(&'static str, Value)> {
    vec![
        ("serial-tool", json!({"function": {"name": "web_search"}})),
        (
            "parallel-tool",
            json!({"function": {"name": "generate_chart"}}),
        ),
        ("nameless", json!({})),
    ]
}

fn chart_cases() -> Vec<(&'static str, Value)> {
    vec![
        (
            "line-with-title",
            json!({"type": "line", "title": "Revenue", "data": [
                {"label": "Q1", "value": 1}, {"label": "Q2", "value": 2.5}]}),
        ),
        ("defaults", json!({"data": [{"label": "a", "value": 1}]})),
        (
            "unknown-type",
            json!({"type": "radar", "data": [{"label": "a", "value": 1}]}),
        ),
        (
            "string-numbers",
            json!({"data": [{"label": "a", "value": "3"}]}),
        ),
        ("drops-and-caps", {
            let mut data = vec![
                json!("nope"),
                json!({"value": 1}),
                json!({"label": "nv"}),
                json!({"label": "n", "value": null}),
                json!({"label": "bad", "value": "abc"}),
                json!({"label": "ok", "value": "3"}),
            ];
            for index in 0..20 {
                data.push(json!({"label": format!("p{index}"), "value": index}));
            }
            json!({"data": data})
        }),
        (
            "pipe-in-label",
            json!({"data": [{"label": "a|b", "value": 1}]}),
        ),
        ("empty-list", json!({"data": []})),
        ("not-a-list", json!({"data": "nope"})),
        ("no-valid-points", json!({"data": [{"value": 1}]})),
        (
            "float-rendering",
            json!({"data": [
            {"label": "x", "value": 100}, {"label": "y", "value": -0.0}]}),
        ),
    ]
}

/// `(label, tool_call, policy)` — `policy` is `None` or a permissive policy.
fn dispatch_cases() -> Vec<(&'static str, Value, bool)> {
    vec![
        (
            "unknown-no-policy",
            json!({"function": {"name": "not_a_tool", "arguments": "{}"}}),
            false,
        ),
        (
            "unknown-with-policy",
            json!({"function": {"name": "not_a_tool", "arguments": "{}"}}),
            true,
        ),
        (
            "chart-no-policy",
            json!({"function": {"name": "generate_chart",
                "arguments": "{\"data\": [{\"label\": \"a\", \"value\": 1}]}"}}),
            false,
        ),
        (
            "chart-with-policy",
            json!({"function": {"name": "generate_chart",
                "arguments": "{\"data\": [{\"label\": \"a\", \"value\": 1}]}"}}),
            true,
        ),
        (
            "chart-empty-args",
            json!({"function": {"name": "generate_chart", "arguments": "{}"}}),
            false,
        ),
        (
            "chart-arguments-as-object",
            json!({"function": {"name": "generate_chart",
                "arguments": {"data": [{"label": "a", "value": 2}]}}}),
            false,
        ),
        (
            "transform-number-summary",
            json!({"function": {"name": "data_transform",
                "arguments": "{\"operation\": \"number_summary\", \"input\": \"1 2 3\"}"}}),
            false,
        ),
        (
            "transform-unknown-operation",
            json!({"function": {"name": "data_transform",
                "arguments": "{\"operation\": \"nope\", \"input\": \"x\"}"}}),
            false,
        ),
        (
            "ssrf-denied-with-policy",
            json!({"function": {"name": "fetch_url",
                "arguments": "{\"url\": \"http://169.254.169.254/\"}"}}),
            true,
        ),
        (
            "path-denied-with-policy",
            json!({"function": {"name": "search_files",
                "arguments": "{\"path\": \"../../etc/passwd\"}"}}),
            true,
        ),
    ]
}

fn outcome_label(outcome: &DispatchOutcome) -> &'static str {
    match outcome {
        DispatchOutcome::Executed(_) => "executed",
        DispatchOutcome::Denied(_) => "denied",
        DispatchOutcome::Unsupported(_) => "unsupported",
        DispatchOutcome::Unported { .. } => "unported",
    }
}

/// Mask the engine-specific suffix of a parse-error message, mirroring the
/// Python probe. The prefix is the oracle's own message and stays compared.
fn mask_engine_error(error: &str) -> String {
    for prefix in ["Invalid JSON", "Invalid regex"] {
        if error.starts_with(prefix) {
            return format!("{prefix}: <engine>");
        }
    }
    error.to_string()
}

/// `(label, operation, input, pattern, path, delimiter)`
fn transform_cases() -> Vec<(
    &'static str,
    &'static str,
    String,
    &'static str,
    &'static str,
    &'static str,
)> {
    let oversized = format!("{{\"a\": \"{}\"}}", "x".repeat(5000));
    vec![
        (
            "extract-simple",
            "extract_regex",
            "a1 b2 c3".to_string(),
            r"([a-z])(\d)",
            "",
            ",",
        ),
        (
            "extract-no-groups",
            "extract_regex",
            "xx yy".to_string(),
            "x+",
            "",
            ",",
        ),
        (
            "extract-optional-group",
            "extract_regex",
            "ab a".to_string(),
            r"a(b)?",
            "",
            ",",
        ),
        (
            "extract-missing-pattern",
            "extract_regex",
            "x".to_string(),
            "",
            "",
            ",",
        ),
        (
            "extract-unicode",
            "extract_regex",
            "中文 abc".to_string(),
            "[a-z]+",
            "",
            ",",
        ),
        (
            "json-whole",
            "json_path",
            "{\"a\": {\"b\": [10, 20]}}".to_string(),
            "",
            "$",
            ",",
        ),
        (
            "json-nested",
            "json_path",
            "{\"a\": {\"b\": [10, 20]}}".to_string(),
            "",
            "$.a.b[1]",
            ",",
        ),
        (
            "json-no-prefix",
            "json_path",
            "{\"a\": {\"b\": [10, 20]}}".to_string(),
            "",
            "a.b[0]",
            ",",
        ),
        (
            "json-missing",
            "json_path",
            "{\"a\": 1}".to_string(),
            "",
            "$.nope",
            ",",
        ),
        (
            "json-unsupported",
            "json_path",
            "{\"a\": 1}".to_string(),
            "",
            "$.a[*]",
            ",",
        ),
        (
            "json-out-of-range",
            "json_path",
            "{\"a\": [1]}".to_string(),
            "",
            "$.a[5]",
            ",",
        ),
        (
            "json-invalid",
            "json_path",
            "not json".to_string(),
            "",
            "$",
            ",",
        ),
        ("json-oversized", "json_path", oversized, "", "$.a", ","),
        (
            "csv-basic",
            "csv_summary",
            "name,value\nx,1\ny,2.5\n".to_string(),
            "",
            "",
            ",",
        ),
        (
            "csv-quoted-delimiter",
            "csv_summary",
            "a,b\n\"x,y\",2\n".to_string(),
            "",
            "",
            ",",
        ),
        (
            "csv-doubled-quotes",
            "csv_summary",
            "a\n\"say \"\"hi\"\"\"\n".to_string(),
            "",
            "",
            ",",
        ),
        (
            "csv-multiline-field",
            "csv_summary",
            "a\n\"one\ntwo\"\n".to_string(),
            "",
            "",
            ",",
        ),
        (
            "csv-thousands",
            "csv_summary",
            ",v\n,\"1,000\"\n".to_string(),
            "",
            "",
            ",",
        ),
        ("csv-empty", "csv_summary", String::new(), "", "", ","),
        (
            "csv-header-only",
            "csv_summary",
            "a,b\n".to_string(),
            "",
            "",
            ",",
        ),
        (
            "csv-semicolon",
            "csv_summary",
            "a;b\n1;2\n".to_string(),
            "",
            "",
            ";",
        ),
        (
            "numbers-simple",
            "number_summary",
            "1 2 3 4".to_string(),
            "",
            "",
            ",",
        ),
        (
            "numbers-signed-decimal",
            "number_summary",
            "-1.5 and +2 and .5".to_string(),
            "",
            "",
            ",",
        ),
        (
            "numbers-none",
            "number_summary",
            "no digits here".to_string(),
            "",
            "",
            ",",
        ),
        ("unknown-operation", "nope", "x".to_string(), "", "", ","),
    ]
}

/// `(label, tool_calls, cancelled)`
fn batch_cases() -> Vec<(&'static str, Vec<Value>, bool)> {
    let chart = |label: &str, id: &str| {
        json!({"id": id, "function": {"name": "generate_chart",
            "arguments": format!("{{\"data\": [{{\"label\": \"{label}\", \"value\": 1}}]}}")}})
    };
    vec![
        (
            "two-parallel",
            vec![
                chart("a", "a"),
                json!({"id": "b", "function": {"name": "data_transform",
                    "arguments": "{\"operation\": \"number_summary\", \"input\": \"1 2\"}"}}),
            ],
            false,
        ),
        (
            "mixed-with-unknown",
            vec![
                chart("a", "a"),
                json!({"id": "b", "function": {"name": "not_a_tool", "arguments": "{}"}}),
            ],
            false,
        ),
        (
            "cancelled-from-start",
            vec![chart("a", "a"), chart("b", "b")],
            true,
        ),
        (
            "capped-at-six",
            (0..9)
                .map(|index| chart("a", &format!("id{index}")))
                .collect(),
            false,
        ),
    ]
}

fn main() {
    let mut out = Map::new();

    for (label, value) in parse_cases() {
        let parsed = parse_tool_arguments(Some(&value));
        out.insert(format!("parse::{label}"), Value::Object(parsed));
    }

    for (label, value) in limit_cases() {
        out.insert(
            format!("limit::{label}"),
            json!(safe_limit(Some(&value), 5, 10)),
        );
    }

    for (label, call) in name_cases() {
        out.insert(
            format!("name::{label}"),
            Value::String(tool_call_name(&call)),
        );
    }

    for (label, call) in parallel_cases() {
        out.insert(
            format!("parallel::{label}"),
            Value::Bool(is_parallel_safe_tool(&call)),
        );
    }

    for (label, arguments) in chart_cases() {
        // The oracle calls `generate_chart(type, title, data)` with
        // `str(arguments.get(key) or default)`; this port reads the same fields
        // off the argument map, so the defaults are applied inside.
        let result = generate_chart(&parse_tool_arguments(Some(&arguments)));
        out.insert(
            format!("chart::{label}"),
            match result {
                Ok(value) => json!({"ok": true, "result": value}),
                Err(failure) => {
                    json!({"ok": false, "error": failure.error, "code": failure.code})
                }
            },
        );
    }

    out.insert(
        "markdown::table".to_string(),
        Value::String(chart_markdown_table(&[
            json!({"label": "a|b", "value": 1.0}),
            json!({"label": "c", "value": 2.5}),
        ])),
    );

    let mut serial: Vec<Value> = SERIAL_TOOL_NAMES
        .iter()
        .map(|name| Value::String((*name).to_string()))
        .collect();
    serial.sort_by(|left, right| left.as_str().cmp(&right.as_str()));
    out.insert("serial::names".to_string(), Value::Array(serial));

    for (label, call, use_policy) in dispatch_cases() {
        let name = tool_call_name(&call);
        let arguments = call
            .get("function")
            .and_then(|function| function.get("arguments"))
            .cloned()
            .unwrap_or(Value::Null);

        // `no-policy` means literally no gate, as the oracle's `policy=None`
        // does — not a default-configured one. Passing a policy here would make
        // the unknown-tool case deny instead of reaching the `Unsupported tool:`
        // envelope, which is the difference between the two error paths.
        let mut permissive = ToolPolicy::permissive();
        let outcome = if use_policy {
            dispatch(&name, &arguments, Some(&mut permissive), None)
        } else {
            dispatch(&name, &arguments, None, None)
        };

        out.insert(
            format!("dispatch::{label}"),
            json!({
                "outcome": outcome_label(&outcome),
                "output": outcome.to_output().cloned().unwrap_or(Value::Null),
                // Every compared case either denies before routing or runs the
                // one ported branch, so no other branch can be invoked.
                "branches": Vec::<String>::new(),
            }),
        );
    }

    // --- data_transform -------------------------------------------------------

    for (label, operation, input, pattern, path, delimiter) in transform_cases() {
        let result = data_transform(operation, &input, pattern, path, delimiter);
        out.insert(
            format!("transform::{label}"),
            match result {
                Ok(value) => json!({"ok": true, "result": value}),
                Err(failure) => json!({
                    "ok": false,
                    "error": mask_engine_error(&failure.error),
                    "code": failure.code,
                }),
            },
        );
    }

    // --- execute_tool_calls ---------------------------------------------------

    for (label, calls, cancelled) in batch_cases() {
        let calls: Vec<Value> = calls;
        let should_cancel = move || cancelled;
        let messages = execute_tool_calls(&calls, &should_cancel, &|call| {
            let name = tool_call_name(call);
            let arguments = call
                .get("function")
                .and_then(|function| function.get("arguments"))
                .cloned()
                .unwrap_or(Value::Null);
            dispatch(&name, &arguments, None, None)
        });
        out.insert(
            format!("batch::{label}"),
            json!({
                "messages": messages,
                // Every compared case uses ported branches or an unknown name, so
                // no recorder-visible branch is invoked — matching Python, where
                // those two branches are the real extracted functions.
                "branches": Vec::<String>::new(),
            }),
        );
    }

    let mut encoded =
        serde_json::to_string_pretty(&Value::Object(out)).expect("serialize probe output");
    encoded.push('\n');
    print!("{encoded}");
}
